src/sources/amqp.rs

```diff
@@ -0,0 +1,565 @@
+use crate::{
+    amqp::AMQPConfig,
+    codecs::{Decoder, DecodingConfig},
+    config::{log_schema, Output, SourceConfig, SourceContext},
+    event::{BatchNotifier, BatchStatus},
+    internal_events::{
+        source::{AMQPEventError, AMQPEventReceived},
+        StreamClosedError,
+    },
+    serde::{bool_or_struct, default_decoding, default_framing_message_based},
+    shutdown::ShutdownSignal,
+    SourceSender,
+};
+use async_stream::stream;
+use bytes::Bytes;
+use chrono::{TimeZone, Utc};
+use codecs::decoding::{DeserializerConfig, FramingConfig};
+use futures::{FutureExt, StreamExt};
+use lapin::{acker::Acker, message::Delivery, Channel};
+use snafu::Snafu;
+use std::io::Cursor;
+use tokio_util::codec::FramedRead;
+use vector_common::finalizer::UnorderedFinalizer;
+use vector_config::configurable_component;
+use vector_core::{
+    config::{AcknowledgementsConfig, LogNamespace},
+    event::Event,
+};
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("Could not create AMQP consumer: {}", source))]
+    AMQPCreateError {
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+    #[snafu(display("Could not subscribe to AMQP queue: {}", source))]
+    AMQPSubscribeError { source: lapin::Error },
+}
+
+/// Configuration for the `amqp` source.
+#[configurable_component(source("amqp"))]
+#[derive(Clone, Debug, Derivative)]
+#[serde(deny_unknown_fields)]
+pub struct AMQPSourceConfig {
+    /// The queue.
+    pub(crate) queue: String,
+
+    /// The consumer.
+    pub(crate) consumer: String,
+
+    /// Connection options for AMQP source.
+    pub(crate) connection: AMQPConfig,
+
+    /// The AMQP routing key.
+    #[serde(default = "default_routing_key")]
+    pub(crate) routing_key: String,
+
+    /// The AMQP exchange key.
+    #[serde(default = "default_exchange_key")]
+    pub(crate) exchange_key: String,
+
+    /// The AMQP offset key.
+    #[serde(default = "default_offset_key")]
+    pub(crate) offset_key: String,
+
+    /// The namespace to use. This overrides the global setting.
+    #[serde(default)]
+    pub log_namespace: Option<bool>,
+
+    #[configurable(derived)]
+    #[serde(default = "default_framing_message_based")]
+    #[derivative(Default(value = "default_framing_message_based()"))]
+    pub(crate) framing: FramingConfig,
+
+    #[configurable(derived)]
+    #[serde(default = "default_decoding")]
+    #[derivative(Default(value = "default_decoding()"))]
+    pub(crate) decoding: DeserializerConfig,
+
+    #[configurable(derived)]
+    #[serde(default, deserialize_with = "bool_or_struct")]
+    pub(crate) acknowledgements: AcknowledgementsConfig,
+}
+
+fn default_routing_key() -> String {
+    "routing".into()
+}
+
+fn default_exchange_key() -> String {
+    "exchange".into()
+}
+
+fn default_offset_key() -> String {
+    "offset".into()
+}
+
+impl Default for AMQPSourceConfig {
+    fn default() -> Self {
```

Comment:
I could be wrong but, isn't it possible to make some additional serde default values to `AMQPSourceConfig` and then not have to implement this `default()` function, since you're calling `impl_generate_config_from_default!` below ?